Ajout d'une fonction globale de collecte et d'exports de métriques pour la supervision

ggielly opened the pull request:
Ajout d'un module générique permettant la supervision externe vers des produits tierces : collecte et exportation de métriques.

Ajout du support initial pour Prometheus et StatsD. Ajout d'un export 'null' idiomatique.

ggielly committed:
Ajout de la collecte et de l'exportation globale des métriques. Ajout d'un premier export pour Prometheus.

ggielly committed:
Rajout de fonctionnalité dans le QueueManager

ggielly committed:
Refactorisation pour suivre logique du code existant sur METRICS_EXPORTER

ggielly committed:
Refactorisation : creer un dossier metrics, sortir les exports dans un fichiers disctincts par produit

ggielly committed:
Ajout des durées en secondes dans le nom des variables

ggielly committed:
Code cleanup

ggielly committed:
Rajout du module pour StatsD et début d'un test d'intégration

ggielly committed:
Remove unused files and add license agreements in English and French

ggielly committed:
Nettoyage + montée de la version du crate prometheus en 0.14

ggielly committed:
Fix erreurs de violation dans le module prometheus

ggielly committed:
Refactor de la gestion des erreurs des metriques et améliorations de la vérification des limites dans l'export Prometheus et StatsD

ggielly committed:
Refactor metrics module and enhance Prometheus and StatsD exporters

- Re-export main types and functions in the metrics module for convenience.
- Introduce optional namespace support in PrometheusExporter for better metric organization.
- Add methods to create PrometheusExporter from environment variables.
- Implement dynamic help text generation for metrics in PrometheusExporter.
- Enhance histogram bucket configuration based on metric name for improved observability.
- Add methods to remove individual metrics and clear all metrics in PrometheusExporter.
- Introduce StatsD exporter creation from environment variables.
- Remove deprecated test_statsd.rs file and replace it with integration tests.
- Add comprehensive benchmarks for the metrics system to evaluate performance under load.
- Implement integration tests for basic operations, validation errors, and convenience methods in the metrics system.

jbousquie commented:
Arrf, va falloir que tu m'expliques ce que ça fait et comment ;-)
Normalement, vaut mieux que tu fasses une branche à part et tu déclenches le PR de toute la branche une fois que c'est OK et que ça marche.
Là, je peux pas déjà merger vu qu'il y a des conflits de toutes façons.

ggielly committed:
Merge branch 'main' into Add-metrics

ggielly commented:
Ça devrait compiler avec le merge conflict en moins... En tout cas chez moi ça fonctionne (tm). J’espère.

Pour comment ça fonctionne, je viens de poser un README_metrics.md qui détaille un peu le fonctionnement du module de supervision, comment il récupère les compteurs, etc.

ggielly committed:
Ajout readme metrics + fix merge conflict ?

ggielly committed:
Merge branch 'Add-metrics' of https://github.com/ggielly/whisper_api into Add-metrics

ggielly commented:
# Module de métriques de l'API Whisper

## Architecture générale du système de métriques

Le système de métriques de l'API Whisper est conçu selon une architecture modulaire qui permet de collecter, traiter et exporter des données de performance vers différents systèmes de monitoring. Le module principal se trouve dans `metrics` et utilise un pattern de traits pour supporter plusieurs backends d'export.

Sont supportés les produits suivants :

- Prometheus
- StatsD
- null

En prévision/liste à faire ?

- Grafana
- NewRelic (K8's, complexe, pas de crate, via OpenTelemetry)
- InfluxDB
- Graphite (redondant avec statsd)
- Datadog (nécessite une clef d'API)

## Structure du module de métriques

### Module Principal (`metrics.rs`)

Le cœur du système est la structure `Metrics` qui implémente le trait `MetricsCollector`. Cette structure :

1. Collecte des métriques asynchrones : toutes les opérations de métriques sont non-bloquantes grâce à l'utilisation de *tokio*
2. Thread-safe : utilise `Arc<dyn  MetricsCollector + Send  + Sync>` pour permettre le partage entre threads
3. Gestion d'erreurs robuste : intègre des validations de sécurité pour éviter les débordements numériques

#### Types de métriques supportées

- Compteurs : pour les événements cumulatifs (jobs soumis, complétés, annulés)
- Jauges : pour les valeurs instantanées (taille de la queue, jobs en cours)
- Histogrammes : pour les distributions de temps (durée des transcriptions)

#### Intégration avec l'API Whisper

Le système de métriques s'intègre à plusieurs niveaux dans l'application :

 1. Dans le gestionnaire de queue (`queue_manager.rs`)

`

// Enregistrement de la soumission d'un job
self.metrics.record_job_submitted(&job_model, &job_language).await;

// Mise à jour de la taille de la queue
self.metrics.set_queue_size(safe_size).await;

// Suivi du nombre de jobs en traitement
self.metrics.set_jobs_processing(safe_count).await;

// Enregistrement de la completion d'un job
metrics.record_job_completed(&job.model, &job.language, duration, "success").await;
`
2. Dans l'application principale (`main.rs`)

#### Métriques HTTP pour les requêtes API

- Exposition d'un endpoint `/metrics` pour Prometheus
- Configuration du backend de métriques selon les variables d'environnement

## Les différents modules d'export

### Module Prometheus (`prometheus.rs`)

Le module Prometheus.rs implémente l'export vers le système de monitoring Prometheus.

#### Fonctionnalités

- Registre de métriques : Utilise `prometheus::Registry` pour organiser toutes les métriques
- Types de métriques Prometheus :
- CounterVec : pour les compteurs avec labels (modèle, langue)
- GaugeVec : pour les jauges avec dimensions
- HistogramVec : pour les distributions temporelles avec buckets configurables

#### Métriques exposées

- whisper_jobs_submitted_total : nombre total de jobs soumis
- whisper_jobs_completed_total : nombre total de jobs complétés
- whisper_queue_size : taille actuelle de la queue
- whisper_jobs_processing : nombre de jobs en cours de traitement
- whisper_job_duration_seconds : histogramme des durées de transcription

#### Configuration

Buckets d'histogramme personnalisables via les variables d'environnement
Labels automatiques pour le modèle et la langue
Exposition via endpoint HTTP /metrics

### Module StatsD (`statsd.rs`)

Le module StatsD permet l'envoi de métriques vers des serveurs StatsD compatibles.

#### Protocole

- UDP : envoi non-bloquant des métriques via UDP
- Format StatsD : respect du protocole standard `metric_name:value|type|@sample_rate`
- Types supportés : c (counter), g (gauge), h (histogram)

#### Fonctionnalités

- Envoi asynchrone : utilise `tokio::net::UdpSocket` pour l'envoi non-bloquant
- Gestion d'erreurs : logging des erreurs sans interrompre l'application
- Formatage automatique : conversion des métriques au format StatsD

#### Configuration

Le fichier `STATSD.md` fournit un guide complet d'intégration avec différents backends (Graphite, InfluxDB, DataDog).

### Module Null (`null.rs`)

Le module Null implémente un backend "no-op" pour les métriques.

#### Utilité

- Environnements de test : désactive les métriques sans modifier le code
- Performance : évite la surcharge des métriques en production si non nécessaires
- Développement : permet de développer sans infrastructure de monitoring

#### Implémentation

No op...
  
#### Configuration et sélection du backend

Le système utilise des variables d'environnement pour sélectionner le backend

`
// Configuration automatique basée sur l'environnement
let metrics = match std::env::var("METRICS_BACKEND").as_deref() {
    Ok("prometheus") => Metrics::new_prometheus(),
    Ok("statsd") => Metrics::new_statsd(),
    _ => Metrics::new_null(), // Backend par défaut
};`

1. Avantages du système
    - Modularité : facilite l'ajout de nouveaux backends de métriques
    - Performance : opérations asynchrones non-bloquantes
    - Fiabilité : gestion d'erreurs robuste qui n'impacte pas l'application principale
    - Flexibilité : support de multiples systèmes de monitoring simultanément
    - Observabilité : métriques détaillées sur tous les aspects de l'API

2. Métriques collectées

Le système surveille en temps réel :

    - Débit : nombre de jobs soumis, traités, échoués
    - Latence : temps de traitement par modèle et langue
    - Charge : taille de la queue, jobs actifs
    - Qualité : taux de succès, types d'erreurs
    - Ressources : utilisation selon la configuration de concurrence

jbousquie commented:
Ok,
- est-ce que tu peux sitoplé traduire le README en englishe et je mettrai un lien (ou je réorganiserai le README principal en plusieurs pages) vers ta feature ?
- me dire quand c'est OK après tests (envoi des curl avec des fichiers audio pour qu'il y ait de l'activité)  et surveillance manuelle du CPU et de la RAM pour s'assurer que ça ne part pas en leak fissa

=> quand tu penses que c'est OK et testé, tu me dis et je merge (je suis incapable de tester ta feature ici)
Gardes ta branche ouverte (et je vais la garder aussi sur Github), comme ça si tu dois intervenir encore après coup, ce sera plus simple d'isoler les modifs.
Moi aussi, je passe en mode "branche" stadire que je ne commite plus rien directement sur main. Ce sera plus propre et surtout plus gérable.

ggielly committed:
Fix des trucs que j'ai cassé

ggielly committed:
Add license field to Cargo.toml and improve comments in config_loader and metrics modules

ggielly committed:
Merge branch 'main' into Add-metrics

ggielly committed:
Fix la branche git !#@%

ggielly committed:
Ajout README anglais pour les metriques

jbousquie commented:
> J'ai fait aussi une modif sur
> 
> * WHISPER_API_TIMEOUT = 600
> * WHISPER_API_KEEPALIVE = 600
> 
> qui deviennent :
> 
> * WHISPER_API_TIMEOUT_SECONDS = 600
> * WHISPER_API_KEEPALIVE_SECONDS = 600

C'est une modification dans le fichier de conf ou dans le nom des constantes dans le code ? je n'arrive pas à le localiser.

Si c'est dans le fichier de conf, je préfèrerais sitoplé que tu remettes les anciens noms de paramètres. Je t'explique pourquoi : 
L'API a déjà été livrée à la DINUM et sur le Tchap pour que d'autres l'essaient. Ça veut dire que si on modifie quoi que ce soit dans le nommage ou fonctionnement de la partie publique (les noms des endpoints, des paramètres, de la conf ou le format des réponses), on crée des "breaking changes", c'est à dire des changements qui cassent le fonctionnement d'une version à une autre.
On peut évidemment créer parfois des breaking changes sur une API, par exemple quand on change de version majeure, qu'on résout un problème bloquant ou qu'on amène un changement de comportement de l'application.
Mais là, ce n'est pas le cas. Ça risque juste de créer du support entre les gens testant deux versions différentes ou du rejet chez ceux qui passent d'une version à l'autre.
Alors la bonne pratique, c'est de ne créer aucun breaking change (tout ce qui est public de l'API reste stable et on modifie autant qu'on veut sous le capot) à ce stade, même si on a initialement mal nommé quelque chose. On le documente plutôt que de le renommer. Là, en l'occurence, même si ce n'est pas dans le nom du paramètre initialement, les timeouts d'API sont habituellement exprimés en secondes en pratique. Tu as raison, on peut le préciser, mais dans ce cas, il faut qu'on le mette dans le commentaire qui décrit la variable/paramètre, mais plus qu'on change le nom une fois publié (même s'il aurait pu être mieux).
Jocépa si je suis bien clair...

ggielly commented:
Dans main.rs : 
![image](https://github.com/user-attachments/assets/6a68fad9-ab84-4bdd-81c8-4b97e630585b)

Pigé pour les modifs, là c'est purement interne, pas d'impact sur la config ou les variables exposées.

ggielly committed:
Rajout d'une suite pour tester les métriques de Prometheus et de Statds.

Ajout dans le fichier de config de la possibilité de lancer ou non les métriques, avec le bon backend.

ggielly committed:
Ajout de tests de validité du fichier de configuration. Chaque variable est testée dans le main avant le lancement du serveur.

ggielly committed:
Refactor configuration handling and add command-line options for documentation generation

- Introduced command-line arguments to generate sample configuration and documentation.
- Updated WhisperConfigValidator to validate and load configurations with type safety.
- Created WhisperConfig from validated configuration for compatibility.
- Improved logging for configuration loading and validation processes.

ggielly committed:
Revert "Refactor configuration handling and add command-line options for documentation generation"

This reverts commit 4bc84e1c9ef032d480fb54f26597707aa8200947.

ggielly committed:
Revert "Ajout de tests de validité du fichier de configuration. Chaque variable est testée dans le main avant le lancement du serveur."

This reverts commit 0499f92d5927008f0d30b5e6ffe1ac2226450607.

ggielly committed:
Prometheus : improve race condition handling, optimize performance with DashMap, and enforce resource limits + update de la documentation.

ggielly committed:
Add safety controls for metric removal and enhance registry management

- Introduced `allow_metric_removal` flag to prevent accidental metric removal in production.
- Updated `remove_*` methods to check the removal flag and return errors if removal is disabled.
- Added `set_allow_metric_removal` and `is_metric_removal_allowed` methods for managing removal permissions.
- Enhanced documentation for registry state management and environment variable validation.

ggielly committed:
Pré allocation du DashMap pour mieux gérer la mémoire. Formule de (max_metrics / 3).max(16) pour commencer.

ggielly committed:
- replaced manual check-then-insert patterns with DashMap's atomic entry() API
- fix TOCTOU race in metric counting with atomic CAS loops

ggielly committed:
Enhance metrics system with StatsD integration and comprehensive documentation

- Updated `test_statsd.rs` to handle errors in metric operations.
- Added detailed StatsD integration guide in `METRICS_STATSD.md`.
- Introduced extensive metrics system documentation in `METRICS_SYSTEM_DOCUMENTATION.md`.
- Created README for metrics module in both English and French, outlining architecture, features, and configuration.
- Implemented null metrics backend for testing and development environments.
- Improved error handling and performance optimizations across the metrics system.

ggielly committed:
Merge branch 'main' into Add-metrics

ggielly committed:
Correction de la descente aux enfers gitesque de l'upstream v0.2

ggielly committed:
Rajout de warning pour trouver pourquoi un metrique ne remonte pas. Creation d'un fichier d'explication sur comment tester les metriques avec CURL et PS.
